### PR TITLE
Use Discovery passage retrieval and Bluemix auto-deploy.

### PR DESCRIPTION
**Pull request by markstur**
* Use Discovery passage retrieval instead of RnR.
* Auto-deploy of all 4 services with wbc- prefix so the deploy button creates/configures/runs.
* Note: a Disco environment not ready error causes Bluemix to restart app. Usually succeeds 2nd time (when env is ready).

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -1,21 +1,14 @@
-# Environment variables
-WORKSPACE_ID=
-CONVERSATION_USERNAME=
-CONVERSATION_PASSWORD=
-#Optional cloudant URL for logging
-CLOUDANT_URL=
```
Comment (stevemart) [inline .env]: i guess cloudant was never used?

Comment (stevemart) [inline .env]: i guess cloudant was never used?

File: .env [hunk 2]
```diff
@@ -1,21 +1,14 @@
...
-#If cloudant url is specified a user name and password must be specified to secure the logging endpoints
-LOG_USER=
-LOG_PASS=
-NLU_USERNAME=
-NLU_PASSWORD=
-
-TONE_ANALYZER_USERNAME=
-TONE_ANALYZER_PASSWORD=
-
-RNR_PASSWORD=
```
Comment (stevemart) [inline .env]: great to see the R&R stuff go away as that is the direction we hear from the watson product team is that we should use discovery from now on.

File: README.md [hunk 3]
```diff
@@ -1,149 +1,76 @@
...
 ![Bluemix Deployments](https://deployment-tracker.mybluemix.net/stats/3999122db8b59f04eecad8d229814d83/badge.svg)
 
-# Multi - Domain Enterprise Digital Assistant
-
-In this developer journey we will create a Watson Conversation, NLU & RnR based chatbot for multiple domains.
-that provides users with:
-
-* Answers important FAQs queries from the multiple domains such as across business lines of Loans, Account Opening, Market Place & Digital Banking of Banking industry.
-* Also the chatbot solves transactional queries such as view transactions, balances, pay utility bills etc.
+# Create a banking chatbot with FAQ discovery, anger detection and natural language understanding
```
Comment (stevemart) [inline README.md]: thanks for renaming this, the more specific the better IMO

File: README.md [hunk 4]
```diff
@@ -1,149 +1,76 @@
...
-* Create a chatbot with dialogs using Watson Conversation on NodeJS platform
-* Learn how to best create Retrieve n Rank collection for long tail questions.
-* Get enhanced natural language capability through NLU API
-* Understand the Banking domain for creating a chatbot & using Conversation and Retrieve n Rank in sync.
-
-
-## Technical Architecture - Enterprise Digital Assistant
+* Create a chatbot that converses via a web UI using Watson Conversation and Node.js
+* Use Watson Discovery with passage retrieval to find answers in FAQ documents
+* Use Watson Tone Analyzer to detect emotion
```
Comment (stevemart) [inline README.md]: detect emotion* in a conversation

File: README.md [hunk 5]
```diff
@@ -1,149 +1,76 @@
...
+* Identify entities with Watson Natural Language Understanding
 
 ![](doc/source/images/Banking-Architecture.png)
 
+### With Watson
```
Comment (stevemart) [inline README.md]: thanks for adding this...

File: README.md [hunk 6]
```diff
@@ -1,149 +1,76 @@
...
 
-## Flow
```
Comment (stevemart) [inline README.md]: we want to keep the flow, but we'll need an architecture diagram, i'm not sure the existing one accurately represents what this app does. i'm happy to help make one if someone provides me with a sketch.

File: README.md [hunk 7]
```diff
@@ -1,149 +1,76 @@
...
-
-*Optionally*, to view the conversation dialog select the workspace and choose the
-**Dialog** tab, here's a snippet of the dialog:
-
-![](doc/source/images/dialog.PNG)
-
-7. Configure Watson Retrieve and Rank
-
-Launch the **Watson retrieve-and-rank** tool. Create a **new data cluster**.
+3. To see the app and services created and configured for this journey, use the Bluemix dashboard. The app is named "watson-banking-chatbot" with a unique suffix. The following services are created and easily identified by the "wbc-" prefix:
```
Comment (stevemart) [inline README.md]: s/"watson-banking-chatbot"/`watson-banking-chatbot`

File: README.md [hunk 8]
```diff
@@ -164,63 +91,34 @@ workspace and select **View details**. Save this ID for later.
...
-> Note: Ensure that you have also created a [**Watson Document Conversion**](https://console.ng.bluemix.net/catalog/services/document-conversion) service as well. Since, Watson RnR uses document conversion at the backend.
-
-## 5. Run the application
+### 4. Load the Discovery documents
 
-### If you used the Deploy to Bluemix button...
+Launch the **Watson Discovery** tool. Create a new environment if prompted. Create a new Collection.
 
-If you used ``Deploy to Bluemix``, most of the setup is automatic to start using the App.
+###### Save the COLLECTION_ID for later.
```
Comment (stevemart) [inline README.md]: why is this a sub title?

File: README.md [hunk 9]
```diff
@@ -164,63 +91,34 @@ workspace and select **View details**. Save this ID for later.
...
 
-### If you want to run it locally...
-
-Otherwise as explained above:
-1. Clone the repo
-2. Instantiate/ Configure the above watson services
-2. Install NodeJS runtime or npm.
-3. Start the app and it can be used at localhost:3000
+Add the file documents and questions present under [`data/Retrieve&Rank/`](data/Retrieve&Rank/)
```
Comment (stevemart) [inline README.md]: okay to keep this for now, but let's rename the directory and lose the ampersand, file systems don't like `&`

File: README.md [hunk 10]
```diff
@@ -164,63 +91,34 @@ workspace and select **View details**. Save this ID for later.
...
 
+### 5. Run the app
+1. Install Node.js runtime or npm.
+1. Start the app.
```
Comment (stevemart) [inline README.md]: indicate how to do this?

``node run.js`` ?

File: README.md [hunk 11]
```diff
@@ -164,63 +91,34 @@ workspace and select **View details**. Save this ID for later.
...
+1. Use the chatbot at localhost:3000.
```
Comment (stevemart) [inline README.md]: is it automatically run on port 3000?

File: manifest.yml [hunk 12]
```diff
@@ -1,30 +1,26 @@
----
-declared-services:
-  retrieve-and-rank-service:
-    label: retrieve_and_rank
```
Comment (stevemart) [inline manifest.yml]: yay no more windows carriage returns

**Comment by stevemart** [comment]
looks good, resolve the conflicts but it's definitely a move in the right direction.

File: app.js [hunk 13]
```diff
@@ -39,6 +40,18 @@ var vcapServices = require( 'vcap_services' );
 
 var bankingServices = require('./banking_services');
 
+const DEFAULT_NAME = "watson-banking-chatbot";
+const DISCOVERY_ACTION = "rnr";  // Replaced RnR w/ Discovery but Conversation action is still "rnr".
+// Discovery environment and collection are found by name if ID is not provided.
+const DISCOVERY_ENVIRONMENT_NAME = process.env.DISCOVERY_ENVIRONMENT_NAME || DEFAULT_NAME;
+const DISCOVERY_COLLECTION_NAME = process.env.DISCOVERY_COLLECTION_NAME || DEFAULT_NAME;
+const DISCOVERY_DOCS = [
```
Comment (stevemart) [inline app.js]: no way to loop through the files in a folder instead?

File: app.js [hunk 14]
```diff
@@ -39,6 +40,18 @@ var vcapServices = require( 'vcap_services' );
 
 var bankingServices = require('./banking_services');
 
+const DEFAULT_NAME = "watson-banking-chatbot";
+const DISCOVERY_ACTION = "rnr";  // Replaced RnR w/ Discovery but Conversation action is still "rnr".
```
Comment (stevemart) [inline app.js]: great comment.

File: app.js (hunk 13 above)
Comment (stevemart) [inline app.js]: also, i'm still very confused by the different things in the data folder, let's delete things we don't need.

File: app.js [hunk 15]
```diff
@@ -645,80 +647,461 @@ function checkForLookupRequests(data, callback){
...
+							console.log("Added document:")
+							console.log(data);
+						}
+					});
+				}
+			}
+			else {
+				console.log("Collection is already loaded with docs.");
+				/* For testing:
+				discovery.deleteCollection(params, (err, data) => {
```
Comment (stevemart) [inline app.js]: remove this chunk?

File: .env (hunk 1 above)
Comment (markstur) [inline .env]: I don't see any sign of it in the banking chatbot. Just talk of it in EDA.

File: README.md (hunk 6 above)
Comment (markstur) [inline README.md]: Yeah. The flow needs some work, but I don't want to lose what was already written, so I just moved it.

File: README.md (hunk 9 above)
Comment (markstur) [inline README.md]: Agree. Ampersand in a dir name really bugs me... but it works. Probably will be moving file to "discovery".

File: README.md (hunk 10 above)
Comment (markstur) [inline README.md]: Saurav added some details. Merging in while resolving conflict.

File: README.md (hunk 11 above)
Comment (markstur) [inline README.md]: local default, yes. Bluemix gives it a random host. Can be explained better when the .env is documented.

File: manifest.yml (hunk 12 above)
Comment (markstur) [inline manifest.yml]: well I got rid of some. I'll do a jshint and more cleanup when there isn't an outstanding PR. Too hard to review/merge if I do it now.

File: README.md (hunk 4 above)
Comment (markstur) [inline README.md]: done. looks better.

File: README.md (hunk 7 above)
Comment (markstur) [inline README.md]: done `wbc-` also

File: README.md (hunk 8 above)
Comment (markstur) [inline README.md]: changed to a note. Looks OK. But the manual setup for discovery needs work along with the .env setting. So will revisit.

File: README.md (hunk 5 above)
Comment (sauravraiguru) [inline README.md]: Yes, we had requested for this journey diagram to mould ours into this. Thank Mark

**Comment by sauravraiguru** [comment]
Merging the pull request, will make data folder naming convention change.

**Pull request merged**